Serialize DML execution on QD

kainwen opened the pull request:
In GDD we downgrade the locklevel from ExclusiveLock
to RowExclusiveLock for UPDATE|DELETE statements to
achieve better performance of concurrent DMLs.

However, some DMLs (like splitupdate, update/delete whose
plans contains Motion) cannot be concurrently executed.
This commit uses a new type of Lock to serialize these
DMLs execution on QD.

For more details, please refer:
  1. the comments for the function `pquery.c:dmlConcurrenceControl`
  2. https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/tvI-VQZbtjI

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`

kainwen commented:
This PR should fix the issue https://github.com/greenplum-db/gpdb/issues/5882
and https://github.com/greenplum-db/gpdb/issues/6488
and

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+ * It is used for serializing DMLs on QD and it uses a new type of
+ * lock to achieve this goal.
+ *
+ * More details please refer the gpdb-dev maillist link:
```
s/maillist/mailing list/

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1446,6 +1450,21 @@ PortalRunMulti(Portal portal, bool isTopLevel,
+				pstmt->rtable && pstmt->resultRelations)
+			{
+				Oid relid = getrelid(linitial_int(pstmt->resultRelations),
+									 pstmt->rtable);
```
Which relation do we care about?  Is it always the initial one?

The blank between `Oid` and `relid` is not correct.

kainwen replied:
@zhangjackey  It is always the first one? Right?

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+	SET_LOCKTAG_VIRTUAL_DML(tag, MyDatabaseId, relid);
+	(void) LockAcquire(&tag, lockmode, false, false);
+
+	return;
```
remove this `return` please.

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+static void
+acquireVirtualDMLLock(Oid relid, LOCKMODE lockmode)
+{
+	LOCKTAG tag;
```
blank between `LOCKTAG` and `tag` is not correct.

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+ * c2. normal delete (delete statement whose plan contains no motion)
+ * c3. split-update (update statement on hash cols of hash-distributed table)
+ * c4. update whose plan has motions (like update t1 set c = c + 1 from t2 where ...)
+ * c5. delete whose plan has motions (like delete from t1 using t2 where t1.c = t2.c)
```
When there is the `returning` clause there will be a gather motion to qd, I think this motion is not related to the grouping here, right?  Then we need to explain which motions we are talking about.

In the other side although we provided a link to the mailing list thread it's still better to explain directly here why a motion makes it impossible to run in parallel.

kainwen replied:
@pivotal-ning-yu In the context of this function, the gather motion is alreay removed if it exists.

kainwen replied:
@pivotal-ning-yu Code changed. I will improve the comment as your advice.

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+ *   c2 ---- RowShareLock
+ *   c3 ---- ExclusiveLock
+ *   c4 ---- ShareRowExclusiveLock
+ *   c5 ---- ShareLock
```
Do these lock modes match exactly with above conflict table?  I mean, if there is not a conflict relation between cx & cy in the table then they will not conflict with each other with the chosen modes, right?

kainwen replied:
@pivotal-ning-yu 
Exactly. It is computed by a prolog program in which any pairs not in the conflict table is considered compatible.

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+dmlConcurrenceControl(Plan *plan, Oid relid)
+{
+	LOCKMODE lockmode = NoLock;
+	Relation rel      = NULL;
```
please correct the blank and the alignment.

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+
+	if (IsA(plan, Motion))
+	{
+		plan = plan->lefttree;
```
It's better to use `outerPlan` here.

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+
+	if (IsA(plan, ModifyTable))
+	{
+		CmdType operation = ((ModifyTable*)plan)->operation;
```
blank between variable type and name is not correct.  Also `s/(ModifyTable*)plan/(ModifyTable *) plan/`

pivotal-ning-yu commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1974,122 @@ PortalSetBackoffWeight(Portal portal)
+
+	Assert(IsA(plan, ModifyTable));
+
+	if (((ModifyTable*)plan)->is_splitupdate)
```
It's better to make a temp variable `ModifyTable *mt = (ModifyTable *) node` here to prevent duplicated type conversions.

pivotal-ning-yu replied:
Looks like this function deserve most comments but unfortunately there is no.  Maybe we could copy-paste or move some from other functions.

kainwen replied:
The following code will still have to access (Plan*)'s nmotionnodes field, which make a temp var for modifytable seem superfluous. Will add more comments.

pivotal-ning-yu commented on src/include/storage/lock.h:
```diff
@@ -203,6 +203,7 @@ typedef enum LockTagType
 	 */
 	LOCKTAG_RESOURCE_QUEUE,		/* ID info for resource queue is QUEUE ID */
 	LOCKTAG_USERLOCK,			/* reserved for old contrib/userlock code */
+	LOCKTAG_VIRTUAL_DML,         /* virtualdml lock to serialize dmls on QD */
```
Let's move the new type after `LOCKTAG_ADVISORY`, there is no need to modify the values of existing types.

kainwen replied:
@pivotal-ning-yu See the macro `#define LOCKTAG_LAST_TYPE	LOCKTAG_ADVISORY`

If we do not make `LOCKTAG_ADVISORY` the last locktag type, we have to change two places.

kainwen commented:
ORCA generate split-update plan for any update statement on randomly-distributed table or non-dist key of hash-dist table. See the isssue: https://github.com/greenplum-db/gpdb/issues/6767

So this bug of orca cause that the algorithm in this PR will treat normal update as split-update, which also try to acquire ExclusiveLock in the new type of lock.

So the test cases hang for the pipeline with ORCA. The dev icw pipeline of my own is green with planner.

vraghavan78 commented:
@kainwen Orca always generate split update. That is by design.

hlinnaka commented:
On 22/01/2019 09:06, Venkatesh Raghavan wrote:
> @kainwen Orca always generate split update. That is by design.

Can we change it?

- Heikki

vraghavan78 commented:
Update in Orca is always represented as a delete and then an insert. We pass the gp_segment_id, ctid which is what determines which tuple to delete and where to insert.

vraghavan78 commented:
@hlinnaka @kainwen, @bhuvnesh2703 @asubramanya are looking that how we can change Orca to accommodate this. Thanks for you patience.

ivannovick commented:
@vraghavan78 thx for checking how ORCA could change for this
@hlinnaka do you have any other/more feedback on implementation of this PR?  I want to get a sense if you generally onboard with the approach of this PR, and if not what is your suggestion?
For context removing the TABLE LOCK on update/delete allows us to achieve radically improved OLTP scores compared to GP5 so that is why this is sensitive to see how to resolve.
Thx all

kainwen commented:
> @hlinnaka @kainwen, @bhuvnesh2703 @asubramanya are looking at how we can change Orca to accommodate this. Thanks for you patience.

Thanks so much!

bhuvnesh2703 commented:
@kainwen 
We looked into the test `Test two operations on same table can be concurrent on QD: normal update, normal delete` in `concurrent_dml_partition.sql` which was causing the hang.
The delete statement acquires a RowShareLock on the root partition on QD. When we execute an update concurrently, it requests an ExclusiveLock on the root partition again on the QD even though the update is on a leaf. This is the correct behavior though and we expect it to block until the delete transaction completes. However, considering the new locking behavior, we could change ORCA to not generate plans with split update when not updating the distribution column, which is definitely going to be better.
However, to unblock this PR, you can turn off optimizer for this test, as it will take time to get the enhancements and the current locking behavior is correct.

BC + AS

Eulerizeit commented:
On Wed, Jan 23, 2019 at 8:18 PM Ivan Novick <notifications@github.com> wrote:
>
> @vraghavan78 thx for checking how ORCA could change for this
> @hlinnaka do you have any other/more feedback on implementation of this PR? I want to get a sense if you generally onboard with the approach of this PR, and if not what is your suggestion?
> For context removing the TABLE LOCK on update/delete allows us to achieve radically improved OLTP scores compared to GP5 so that is why this is sensitive to see how to resolve.
> Thx all

My concern is that we are adding a net new lock week(s) before we
beta, testing in the executor code isn't great and this doesn't give
us a lot of time to just have ICW flake out a bunch allowing us to see
if something is weird.

For the single use case of Split update, I would be much more
comfortable seeing this in 7 and in 6 moving the locking back to the
5.x level.

-- Rob

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

kainwen commented:
> .

Thanks so much. Not only `concurrent_dml_partition.sql`, but also `concurrent_dml` has the same problem.

This PR needs more reviews.

kainwen commented:
> @kainwen
> We looked into the test `Test two operations on same table can be concurrent on QD: normal update, normal delete` in `concurrent_dml_partition.sql` which was causing the hang.
> The delete statement acquires a RowShareLock on the root partition on QD. When we execute an update concurrently, it requests an ExclusiveLock on the root partition again on the QD even though the update is on a leaf. This is the correct behavior though and we expect it to block until the delete transaction completes. However, considering the new locking behavior, we could change ORCA to not generate plans with split update when not updating the distribution column, which is definitely going to be better.
> However, to unblock this PR, you can turn off optimizer for this test, as it will take time to get the enhancements and the current locking behavior is correct.
> 
> BC + AS

Thanks so much. Not only `concurrent_dml_partition.sql`, but also `concurrent_dml` has the same problem.

This PR needs more reviews.

bhuvnesh2703 commented:
@kainwen 

> Not only concurrent_dml_partition.sql, but also concurrent_dml has the same problem.
Yes, we tried that too.

pengzhout commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,157 @@ PortalSetBackoffWeight(Portal portal)
+
+	if (commandType == CMD_UPDATE)
+	{
+		if (plan->nMotionNodes > 0)
```
Hmm, nMotionNodes also contains the motion within initplan which should not bother the concurrent update.

If we can exclude init plan case, it would be better. without it also be acceptable.

kainwen replied:
OK. Let me consider the issue and maybe add more test cases. Thanks.

kainwen replied:
We cannot exclude initplan case. See the code here: https://github.com/greenplum-db/gpdb/blob/master/src/backend/executor/execMain.c#L4103

InitPlan's motion will also lead to a crash in EPQ.

kainwen commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,157 @@ PortalSetBackoffWeight(Portal portal)
+		plan = outerPlan(plan);
+		if (plan == NULL)
+			return;
+	}
```
OK. I will replace the code here with an assert. Will be an assert superfluous?

pengzhout replied:
Please put the "Assert(IsA(plan, ModifyTable) || IsA(plan, DML));" into deduceLockMode(),  so the duplicated if block here can be removed.

kainwen replied:
Thanks. Will improve the code.

kainwen commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,157 @@ PortalSetBackoffWeight(Portal portal)
+ *   c5 ---- ShareLock
+ */
+static void
+dmlConcurrenceControl(PlannedStmt *pstmt, CmdType commandType, Oid relid)
```
OK. Will improve the code.

kainwen commented on src/backend/tcop/pquery.c:
```diff
@@ -1446,6 +1459,26 @@ PortalRunMulti(Portal portal, bool isTopLevel,
+				 pstmt->commandType == CMD_DELETE) &&
+				pstmt->rtable && pstmt->resultRelations)
+			{
+				Oid relid = getrelid(linitial_int(pstmt->resultRelations),
```
Inherit table has been taken into consideration already and there is a test script for it. 
@pengzhout What is the problem?

pengzhout commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,157 @@ PortalSetBackoffWeight(Portal portal)
+	 * AO table is already opened using ExclusiveLock at
+	 * parser stage, so here we can just skip them.
+	 */
+	rel = heap_open(relid, NoLock);
```
add FIXME or TODO? now we have two methods to do GPDB-specified DML concurrent control, it is easy to confuse, as here shows, we need to check if the DML operation has been controlled by another method. Actually, we don't even need to explicitly skip the AO table here, acquiring a lock on virtual lock doing no harm and code can be cleaner and clearer.

kainwen replied:
OK. A TODO is added.

hlinnaka commented:
> @hlinnaka do you have any other/more feedback on implementation of this PR? I want to get a sense if you generally onboard with the approach of this PR, and if not what is your suggestion?

I don't like adding a new lock type for this. However, I don't have any better suggestions :-(. I think it's the least bad option we have available, and we just have to suck it up.

The ideal solution would be to change the executor so that it would perform the deletion first, and do the Insertion only for those tuples that were successfully deleted. That would require quite big changes in the executor though. And we'd also need to modify AO tables, because they don't have a 'ctid' pointer today. That's an even bigger change. So that option is not feasible, certainly not for GPDB 6, probably not for GPDB 7 either. Even if we fixed some of the problematic cases - which would be nice on its own - we'd still need to have a mechanism like this for the remaining cases.

Speaking of partial solutions, though: For heap tables, this is only an issue in READ COMMITTED mode. We could skip the new lock in REPEATABLE READ mode, because in REPEATABLE READ, you get a serialization error, if you try to modify a row that's been concurrently modified by another transaction.

hlinnaka commented on src/backend/tcop/pquery.c:
```diff
@@ -1446,6 +1459,26 @@ PortalRunMulti(Portal portal, bool isTopLevel,
+				relid = root == InvalidOid ? relid : root;
+
+				if (relid != InvalidOid)
+					dmlConcurrenceControl(pstmt, relid);
```
This code seems misplaced. The target relations are locked in InitPlan(), we should acquire these new locks there, as well.

kainwen replied:
We have tried to invoke this code in InitPlan at first but our cases failed. But that is late because GetSnapshot is invoked before InitPlan(). 

Why before GetSnapshot is important? 

Suppose two transactions txn1, txn2 and suppose that we put the concurrence control in InitPlan.

txn1 executes:
`begin; update t set c = 999 where c = 1; end;`

txn2 executes:
`begin; update t set c = 888 from t2 where t.c = t2.c end;` -- this update contains motions

1. txn1 holds the new type of lock at InitPlan
2. txn2 will wait on the new type of lock at InitPlan
3. as txn1, txn2 is in my snapshot->xip array, means to me, txn2 is in progress
4. as txn2, tnx1 is in my snapshot->xip array, means to me, txn1 is in progress
5. txn1 commited
6. txn2 continue, (why txn2 can continue? because txn1 is committed, but this information `txn1 has been commited` is not known to txn2). 
7. txn2 try to update the same tuple updated by txn1. It sees the xmax in that tuple is in its snapshot->xip, so it tries to wait this xid, and the wait will be over immediately.
8. txn2 finds the result is HeapUpdated, so try to do a EvalPlanQual on QE
9. txn2 fails

hlinnaka commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,164 @@ PortalSetBackoffWeight(Portal portal)
 }
+
+static void
+acquireVirtualDMLLock(Oid relid, LOCKMODE lockmode)
```
I think this belongs in `lmgr.c`, where we have other functions like this.

kainwen replied:
Thanks, will improve the code as your advice.

hlinnaka commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,164 @@ PortalSetBackoffWeight(Portal portal)
+/*
+ * dmlConcurrenceControl
+ *
+ * This function can only be called on QD and before Getsnapshot.
```
Why does it need to be called before "Getsnapshot"?

hlinnaka commented on src/backend/tcop/pquery.c:
```diff
@@ -1955,3 +1988,164 @@ PortalSetBackoffWeight(Portal portal)
+ *
+ * Before invoking this function, it has stripped off the top motion
+ * node, which is gather motion for DML or Modifytable plan (because
+ * of returning clause).
```
This is really complicated.

The executor should not have to reverse-engineer the plan, to figure out what locks are required. The planner should record that information in the PlannedStmt. 

PlannedStmt.resultRelations stores the information on which tables are being modified. InitPlan() uses that, to acquire appropriate locks on them. In addition to that, there's PlannedStmt.rowMarks list. We're already using that to store information on which tables need to be locked, see ROW_MARK_TABLE_SHARE and ROW_MARK_TABLE_EXCLUSIVE. Let's use the same mechanism for this.

hlinnaka commented on src/backend/utils/adt/lockfuncs.c:
```diff
@@ -34,6 +34,7 @@ static const char *const LockTagTypeNames[] = {
 	"virtualxid",
 	"append-only segment file",
 	"object",
+	"virtual-dml",
```
I don't like this name. What is "virtual" about it? (In "virtualxid", the lock key is the virtual transaction ID, hence the name).

The lock is taken on a relation, so I'd suggest "relation-dml" or something like that.

kainwen replied:
@hlinnaka 
Thanks, I will improve the code as the advice.

kainwen commented:
@hlinnaka Thanks for your comments.

> > @hlinnaka do you have any other/more feedback on implementation of this PR? I want to get a sense if you generally onboard with the approach of this PR, and if not what is your suggestion?
> 
> I don't like adding a new lock type for this. However, I don't have any better suggestions :-(. I think it's the least bad option we have available, and we just have to suck it up.
> 

The problem is hard in MPP-database and Postgres never face such situation: we have to decide concurrence control based on plan(split-update? has motion?).

> The ideal solution would be to change the executor so that it would perform the deletion first, and do the Insertion only for those tuples that were successfully deleted. That would require quite big changes in the executor though. And we'd also need to modify AO tables, because they don't have a 'ctid' pointer today. That's an even bigger change. So that option is not feasible, certainly not for GPDB 6, probably not for GPDB 7 either. Even if we fixed some of the problematic cases - which would be nice on its own - we'd still need to have a mechanism like this for the remaining cases.
> 

Can this solution handle the EvalPlanQual issuse? Greenplum now cannot execute EvalPlanQual on QE if the subplan contains motions. This is the hardest issue need to solve compared to split-update.

> Speaking of partial solutions, though: For heap tables, this is only an issue in READ COMMITTED mode. We could skip the new lock in REPEATABLE READ mode, because in REPEATABLE READ, you get a serialization error, if you try to modify a row that's been concurrently modified by another transaction.

Thanks for the advice, will improve the code.

hlinnaka commented:
On 24/01/2019 10:41, Zhenghua Lyu wrote:
>> The ideal solution would be to change the executor so that it would
>> perform the deletion first, and do the Insertion only for those
>> tuples that were successfully deleted. That would require quite big
>> changes in the executor though. And we'd also need to modify AO
>> tables, because they don't have a 'ctid' pointer today. That's an
>> even bigger change. So that option is not feasible, certainly not
>> for GPDB 6, probably not for GPDB 7 either. Even if we fixed some
>> of the problematic cases - which would be nice on its own - we'd
>> still need to have a mechanism like this for the remaining cases.
> 
> Can this solution handle the EvalPlanQual issuse? Greenplum now
> cannot execute EvalPlanQual on QE if the subplan contains motions.
> This is the hardest issue need to solve compared to split-update.

Hmm, I guess it wouldn't. But neither does this patch, AFAICS. You could
have a ModifyTable node with a subplan with motions, even without a 
Split Update.

- Heikki

kainwen commented:
> On 24/01/2019 10:41, Zhenghua Lyu wrote: > The ideal solution would be to change the executor so that it would > perform the deletion first, and do the Insertion only for those > tuples that were successfully deleted. That would require quite big > changes in the executor though. And we'd also need to modify AO > tables, because they don't have a 'ctid' pointer today. That's an > even bigger change. So that option is not feasible, certainly not > for GPDB 6, probably not for GPDB 7 either. Even if we fixed some > of the problematic cases - which would be nice on its own - we'd > still need to have a mechanism like this for the remaining cases. Can this solution handle the EvalPlanQual issuse? Greenplum now cannot execute EvalPlanQual on QE if the subplan contains motions. This is the hardest issue need to solve compared to split-update.
> Hmm, I guess it wouldn't. But neither does this patch, AFAICS. You could have a ModifyTable node with a subplan with motions, even without a Split Update. - Heikki

Hi, @hlinnaka 

This patch fixes two problems:
* splitupdate
* evalplanqual

We have test cases for that. The modified code is just less then 200 lines. But we provide about thousands of lines of test cases to test each case. 

```
Insert is OK and other DMLs can be grouped into five class:

c1. normal update (update statement whose plan contains no motion)
c2. normal delete (delete statement whose plan contains no motion)
c3. split-update (update statement on hash cols of hash-distributed table)
c4. update whose plan has motions (like update t1 set c = c + 1 from t2 where ...)
c5. delete whose plan has motions (like delete from t1 using t2 where t1.c = t2.c)
If these operations are on the same table, the conflict relation among them
should be:

c1 conflict with [c3, c4, c5]
c2 conflict with [c3]
c3 conflict with [c1, c2, c3, c4, c5]
c4 conflict with [c1, c3, c4, c5]
c5 conflict with [c1, c3, c4]
Explanation on the conflict relations:

Splitupdate has to be conflict with any other operations.
This is because splitupdate's ExecInsert cannot be blocked
For the remaining operations, for each pair that might lead
to EvalPlanQual a subplan with motions on QE, such pair has
to be conflict with each other. And to invoke EvalPlanQual
there is a prerequisite: the tuple is marked as HeapUpdated
by some other transactions using heap_update.
So the lockmode for these five operations can be:

c1 ---- RowExclusiveLock
c2 ---- RowShareLock
c3 ---- ExclusiveLock
c4 ---- ShareRowExclusiveLock
c5 ---- ShareLock

```

kainwen commented on src/test/isolation2/expected/concurrent_dml.out:
```diff
@@ -0,0 +1,1397 @@
+drop table t_concurrent_dmls_auxiliary;
+DROP
+
+-- Test two operations on same table should be serialized on QD: normal update, delete whose plan contains motion
```
Hi @hlinnaka 

I add comments before each case to describe what it is fore.

Splitupdate is just a small fraction of all the test cases.

mkiyama commented:
@kainwen 
I have a question. 
Does this PR add a new lock mode or does it change an existing lock mode?
Can you describe the change?

thanks

Link to the docs for LOCK with information the current lock modes.
http://docs-gpdb-6x.cfapps.io/600/ref_guide/sql_commands/LOCK.html

kainwen commented:
> So, there is another problem, which, admittedly, was also a problem in 5.
> With Orca on, on this branch the following sequence of events will produce wrong results:
> 
> ```
> create table t1 (a int, b int) distributed by (a) partition by range(b) (start(1) end(5) every(2));
> insert into t1 values(1,4);
> ```
> S1 = Session 1
> S2 = Session 2
> 
> ```
> S1: begin;
> S2: begin;
> S1: update t1 set b = b - 2;
> S2: delete from t1_1_prt_2 where b = 4;
> S1: commit;
> S2: commit;
> ```
> Now, if you query t1, the tuple which should have been deleted is still there but is updated.
> 
> Planner actually throws an error because we are trying to move a tuple from one partition to another in updating the partition key.
> 
> This is the lock situation after executing both commands in each session before committing either (of course the delete is waiting until we commit S1).
> 
> ```
> S1:
> postgres=# select relation::regclass, locktype, mode, granted, mppsessionid, gp_segment_id from pg_locks where mppsessionid = 7 order by relation;
>   relation  |   locktype    |       mode       | granted | mppsessionid | gp_segment_id
> ------------+---------------+------------------+---------+--------------+---------------
>  t1         | relation      | RowExclusiveLock | t       |            7 |             2
>  t1         | virtual-dml   | ExclusiveLock    | t       |            7 |            -1
>  t1         | relation      | RowExclusiveLock | t       |            7 |             1
>  t1         | relation      | RowExclusiveLock | t       |            7 |            -1
>  t1         | relation      | RowExclusiveLock | t       |            7 |             0
>  t1_1_prt_1 | relation      | AccessShareLock  | t       |            7 |             0
>  t1_1_prt_1 | relation      | RowExclusiveLock | t       |            7 |             2
>  t1_1_prt_1 | relation      | AccessShareLock  | t       |            7 |             2
>  t1_1_prt_1 | relation      | AccessShareLock  | t       |            7 |             1
>  t1_1_prt_2 | relation      | AccessShareLock  | t       |            7 |             0
>  t1_1_prt_2 | relation      | AccessShareLock  | t       |            7 |             1
>  t1_1_prt_2 | relation      | RowExclusiveLock | t       |            7 |             2
>  t1_1_prt_2 | relation      | AccessShareLock  | t       |            7 |             2
> 
>  S2:
>  postgres=# select relation::regclass, locktype, mode, granted, mppsessionid, gp_segment_id from pg_locks where mppsessionid = 9 order by relation;
>   relation  |  locktype   |       mode       | granted | mppsessionid | gp_segment_id
> ------------+-------------+------------------+---------+--------------+---------------
>  t1         | virtual-dml | RowShareLock     | f       |            9 |            -1
>  t1_1_prt_2 | relation    | RowExclusiveLock | t       |            9 |            -1
> ```
> Now, I spent some time trying to come up with a case where the fact that we only take a RowExclusiveLock on the exact leaf partition on the particular segment which we are updating would cause a problem with planner, however, I have not yet come up with something. It feels like that might be a problem, but I have no proof.

I have created a story in MPP's tracker to check the problem, thanks!
